exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/dml/MergeTableTest.kt

```diff
@@ -0,0 +1,489 @@
+                    it[Dest.key] = Source.key
+                    it[Dest.value] = Dest.value + 2
+                }
+                updateWhenMatched(and = (Source.value less 12)) {
+                    it[Dest.key] = Source.key
+                    it[Dest.value] = Dest.value + 3
+                    it[Dest.optional] = "never"
+                }
+            }
+
+            assertEquals(101, Dest.getByKey("test1")[Dest.value])
+            assertEquals(102, Dest.getByKey("test2")[Dest.value])
+            assertEquals(102, Dest.getByKey("test3")[Dest.value])
+        }
+    }
+
+    @Test
+    fun testMultipleDeletes() {
+        withMergeTestsTables(excludeSettings = TestDB.oracleRelatedDB + TestDB.sqlServerRelatedDB) {
+            Insert.source("test1", 5)
+            Insert.source("test2", 10)
+            Insert.source("test3", 15)
+
+            Insert.dest("test1", 100)
+            Insert.dest("test2", 100)
+            Insert.dest("test3", 100)
+
+            Dest.mergeFrom(Source, on = { Source.key eq Dest.key }) {
+                deleteWhenMatched(and = (Source.value less 7))
+                deleteWhenMatched(and = (Source.value greater 12))
+            }
+
+            assertNull(Dest.getByKeyOrNull("test1"))
+            assertNotNull(Dest.getByKeyOrNull("test2"))
+            assertNull(Dest.getByKeyOrNull("test3"))
+        }
+    }
+
+    @Test
+    fun testMultipleInsertsInOracleAndSqlserver() {
+        withMergeTestsTables(excludeSettings = allDbExcept(TestDB.oracleRelatedDB + TestDB.sqlServerRelatedDB)) {
+            assertFailsWith(IllegalArgumentException::class) {
+                Dest.mergeFrom(Source, on = { Source.key eq Dest.key }) {
+                    insertWhenNotMatched {
+                        it[Dest.key] = Source.key
+                        it[Dest.value] = Source.value
+                    }
+                    insertWhenNotMatched {
+                        it[Dest.key] = Source.key
+                        it[Dest.value] = Source.value
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    fun testMultipleUpdatesInOracleAndSqlserver() {
+        withMergeTestsTables(excludeSettings = allDbExcept(TestDB.oracleRelatedDB + TestDB.sqlServerRelatedDB)) {
+            assertFailsWith(IllegalArgumentException::class) {
+                Dest.mergeFrom(Source, on = { Source.key eq Dest.key }) {
+                    updateWhenMatched {
+                        it[Dest.value] = Source.value
+                    }
+                    updateWhenMatched {
+                        it[Dest.value] = Source.value
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    fun testDeleteInOracleAndSqlserver() {
+        withMergeTestsTables(excludeSettings = allDbExcept(TestDB.oracleRelatedDB + TestDB.sqlServerRelatedDB)) {
+            assertFailsWith(IllegalArgumentException::class) {
```

Comment:
You don't have to use it, just letting you know that there's a special test function for this `expectException()` that may be useful. Maybe less useful now though since running tests are helpfully labelled with the dialect name.